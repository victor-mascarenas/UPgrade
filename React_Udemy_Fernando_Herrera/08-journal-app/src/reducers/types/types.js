export const types = {
    LOGIN: '[Auth] Login',
    LOGOUT: '[Auth] Logout'
}