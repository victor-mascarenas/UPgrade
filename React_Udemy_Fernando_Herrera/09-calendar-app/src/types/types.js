export const types = {
    UI_OPEN_MODAL: '[UI] Open Modal',
    UI_CLOSE_MODAL: '[UI] Close Modal'
};