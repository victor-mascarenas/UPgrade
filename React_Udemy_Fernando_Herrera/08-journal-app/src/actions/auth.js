import { types } from "../reducers/types/types";
import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { endLoading, startLoading } from "./ui";
import Swal from 'sweetalert2';
import { notesLogout } from "./notes";

export const startLogin = (email, password) => {
    return (dispatch) => {//dispatch comes from Thunk
        dispatch(startLoading());
        return firebase.auth().signInWithEmailAndPassword(email, password)
            .then(({user}) => {
                dispatch(login(user.uid, user.displayName));
            })
            .catch(error => {
                Swal.fire('Error', error.message, 'error');
            })
            .finally(() => {
                dispatch(endLoading());
            });
    };
};

export const startCredentialsRegistration = (email, password, name) => {
    return (dispatch) => {
        firebase.auth().createUserWithEmailAndPassword(email, password)
            .then(async ({user}) => {
                await user.updateProfile({
                    displayName: name
                });
                dispatch(login(user.uid, user.displayName));
            })
            .catch(error => {
                Swal.fire('Error', error.message, 'error');
            });
    };
};

export const startGoogleLogin = () => {
    return (dispatch) => {
        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({user}) => {
                dispatch(login(user.uid, user.displayName));
            })
            .catch((error) => {
                Swal.fire('Error', error.message, 'error');
            });
    };
};

export const login = (uid, displayName) => (
    {
        type: types.LOGIN,
        payload: {
            uid,
            displayName
        }
    }
);

export const startLogout = () => {
    return async (dispatch) => {
        await firebase.auth().signOut();
        dispatch(logout());
        dispatch(notesLogout());
    };
};

export const logout = () => {
    return {
        type: types.LOGOUT
    };
}