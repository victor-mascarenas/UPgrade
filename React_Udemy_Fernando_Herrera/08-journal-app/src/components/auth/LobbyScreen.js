import React from 'react'

export const LobbyScreen = () => {
    return (
        <div className='auth__main'>
            <div className='auth__box-container text-center'>
                <b>Checking your status</b>
                <br/>
                Please wait...
            </div>
        </div>
    )
}
