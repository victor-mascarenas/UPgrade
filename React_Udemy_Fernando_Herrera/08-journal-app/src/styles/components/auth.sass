.auth__main
    align-items: center
    background-color: $primary
    display: flex
    justify-content: center
    height: 100vh
    margin: 0px
    width: 100vw
.auth__box-container
    background-color: white
    border-radius: 2px
    box-shadow: 0px 3px $dark-gray
    padding: 20px
    width: 250px