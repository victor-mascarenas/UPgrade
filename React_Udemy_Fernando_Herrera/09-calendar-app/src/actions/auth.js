import Swal from "sweetalert2";
import { fetchWithoutToken } from "../helpers/fetch";
import { types } from "../types/types";

const storeToken = (token) => {
    localStorage.setItem('token', token);
    localStorage.setItem('token-start', new Date().getTime());
};

export const startLogin = (email, password) => {
    return async (dispatch) => {
        const res = await fetchWithoutToken('auth', {
            email,
            password
        }, 'POST');
        const body = await res.json();
        if (body.ok) {
            storeToken(body.token);
            dispatch(login({
                uid: body.uid,
                name: body.name
            }));
        } else {
            Swal.fire('Error', body.msg, 'error');
        }
    };
};
export const startRegister = (email, password, name) => {
    return async (dispatch) => {
        const res = await fetchWithoutToken('auth/new', {
            name,
            email,
            password
        }, 'POST');
        const body = await res.json();
        if (body.ok) {
            storeToken(body.token);
            dispatch(login({
                uid: body.uid,
                name: body.name
            }));
        } else {
            Swal.fire('Error', body.msg, 'error');
        }
    };
};

const login = (user) => ({
    type: types.AUTH_LOGIN,
    payload: user
});