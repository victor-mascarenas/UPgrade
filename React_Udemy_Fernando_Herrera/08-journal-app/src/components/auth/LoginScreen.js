import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import { startGoogleLogin, startLogin } from '../../actions/auth';
import useForm from '../../hooks/useForm';

export const LoginScreen = () => {
    const {values: formValues, inputOnChange} = useForm({
        email: 'nando@gmail.com',
        password: '123456'
    });
    const dispatch = useDispatch();
    const state = useSelector(state => state);

    const {email, password} = formValues;

    const loginSubmit = (e) => {
        e.preventDefault();
        dispatch(startLogin(email, password));//async action
    };
    const handleGoogleLogin = () => {
        dispatch(startGoogleLogin());
    };

    return (
        <>
            <h3 className='auth__title'>Login</h3>
            <form onSubmit={loginSubmit} className='animate__animated animate__fadeIn'>
                <input className='auth__input' type='text' placeholder='Email' name='email' autoComplete='off' value={email} onChange={inputOnChange}/>
                <input className='auth__input' type='password' placeholder='Password' name='password' value={password} onChange={inputOnChange}/>
                <button className='btn btn-primary btn-block' type='submit' disabled={state.ui.loading ? true : false}>Login</button>
                <div className='auth__social-networks'>
                    <p>Login woth social networks</p>
                    <div className="google-btn" onClick={handleGoogleLogin}>
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>
                <Link className='link' to='/auth/register'>Create an account instead</Link>
            </form>
        </>
    )
}
