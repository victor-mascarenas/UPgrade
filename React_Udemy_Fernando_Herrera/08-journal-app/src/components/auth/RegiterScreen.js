import React from 'react'
import { Link } from 'react-router-dom'
import useForm from '../../hooks/useForm';
import validator from 'validator';
import { useDispatch } from 'react-redux';
import { removeError, setError } from '../../actions/ui';

export const RegiterScreen = () => {
    const {values: formValues, inputOnChange} = useForm({
        name: 'Victor',
        email: 'nando@gmail.com',
        password: '123456',
        confirmPassword: '123456'
    });
    const dispatch = useDispatch();
    const {name, email, password, confirmPassword} = formValues;

    const handleRegister = (e) => {
        e.preventDefault();
        if (isFormValid()) {
            console.log('correcto');
        }
    };

    const dispatchError = (msg, set = true) => {
        if (set) {
            dispatch(setError(msg));
        } else {
            dispatch(removeError());
        }
    };

    const isFormValid = () => {
        let valid = false;
        if (name.trim().length === 0) {
            dispatchError('Name is required');
        } else if (email.trim().length === 0) {
            dispatchError('Email is required');
        } else if (!validator.isEmail(email)) {
            dispatchError('Email is not valid');
        } else if (password !== confirmPassword) {
            dispatchError('Passwords are ot the same');
        } else if (password.trim().length < 5) {
            dispatchError('Password needs to be at least 5 characters long');
        } else {
            dispatchError('', false);
            valid = true;
        }
        return valid;
    };

    return (
        <>
            <h3 className='auth__title'>Create an account</h3>
            <form onSubmit={handleRegister}>
                {
                    <div className='auth__alert-error'>
                        Campos no validos
                    </div>
                }
                <input className='auth__input' type='text' placeholder='Name' name='name' autoComplete='off' value={name} onChange={inputOnChange}/>
                <input className='auth__input' type='text' placeholder='Email' name='email' autoComplete='off' value={email} onChange={inputOnChange}/>
                <input className='auth__input' type='password' placeholder='Password' name='password' value={password} onChange={inputOnChange}/>
                <input className='auth__input' type='password' placeholder='Repeat password' name='confirmPassword' value={confirmPassword} onChange={inputOnChange}/>
                <button className='btn btn-primary btn-block mb-5' type='submit'>Sign up</button>
                <Link className='link' to='/auth/login'>Login instead</Link>
            </form>
        </>
    )
}
