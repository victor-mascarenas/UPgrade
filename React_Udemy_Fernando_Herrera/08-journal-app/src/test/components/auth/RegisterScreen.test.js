import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { MemoryRouter } from 'react-router-dom';
import { RegiterScreen } from '../../../components/auth/RegiterScreen';
import { types } from '../../../reducers/types/types';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {
    auth: {
        uid: 'TESTING',
        displayName: null
    },
    ui: {
        loading: false,
        errorMsg: ''
    }
};
const store = mockStore(initState);

const wrapper = mount(
    <Provider store={store}>
        <MemoryRouter>
            <RegiterScreen/>
        </MemoryRouter>
    </Provider>
);

describe('Pruebas en RegisterScreen', () => {
    beforeEach(() => {
        store.clearActions();
    });

    test('Debe mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });
    test('Debe hacer el dispatch de la accion respectiva', () => {
        const emailField = wrapper.find('input[name="email"]');
        emailField.simulate('change', {
            target: {
                value: '',
                name: 'email'
            }
        });
        wrapper.find('form').simulate('submit', {
            preventDefault: jest.fn()
        });
        const actions = store.getActions();
        expect(actions[0]).toEqual({
            type: types.UI_SET_ERROR,
            payload: {
                errorMsg: 'Email is required'
            }
        });
    });
});