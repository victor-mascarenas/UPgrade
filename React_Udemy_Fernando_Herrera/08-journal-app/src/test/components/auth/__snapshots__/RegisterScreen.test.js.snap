// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en RegisterScreen Debe mostrarse correctamente 1`] = `
Array [
  <h3
    className="auth__title"
  >
    Create an account
  </h3>,
  <form
    className="animate__animated animate__fadeIn"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="Victor"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value="nando@gmail.com"
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="123456"
    />
    <input
      className="auth__input"
      name="confirmPassword"
      onChange={[Function]}
      placeholder="Repeat password"
      type="password"
      value="123456"
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Sign up
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Login instead
    </a>
  </form>,
]
`;
